README.md
react-native-image-zoom-viewer 在2021年9月28号已停更，这个为 fork 版本，修复了一些 bug 以及少许更新。

## Show Cases

**Swiper image**

![](https://cloud.githubusercontent.com/assets/7970947/21086300/388dedfc-c056-11e6-955e-0a2a0b541f7f.gif)

**Zoom while sliding**

![](https://cloud.githubusercontent.com/assets/7970947/21086323/7355face-c056-11e6-8d68-384000d41d47.gif)

**Swipe down**

![](https://user-images.githubusercontent.com/7970947/37651584-8b642458-2c73-11e8-9ecc-ef7c72aca1be.gif)

## Getting Started

### Installation

```bash
npm i react-native-image-zoom-viewer-1
```

### Basic Usage

- Install create-react-native-app first

```bash
$ npm install -g create-react-native-app
```

- Initialization of a react-native project

```bash
$ create-react-native-app AwesomeProject
```

- Then, edit `AwesomeProject/App.js`, like this:

```typescript
import { Modal } from 'react-native';
import ImageViewer from 'react-native-image-zoom-viewer-1';

const images = [{
    // Simplest usage.
    url: 'https://avatars2.githubusercontent.com/u/7970947?v=3&s=460',

    // width: number
    // height: number
    // Optional, if you know the image size, you can set the optimization performance

    // You can pass props to <Image />.
    props: {
        // headers: ...
    }
}, {
    url: '',
    props: {
        // Or you can set source directory.
        source: require('../background.png')
    }
}]

export default class App extends React.Component {
    render: function() {
        return (
            <Modal visible={true} transparent={true}>
                <ImageViewer imageUrls={images}/>
            </Modal>
        )
    }
}
```

### Props

| parameter              | type                                                                                   | required | description                                                                                                                                                                                                                          | default                                                   |
| :--------------------- | :------------------------------------------------------------------------------------- | :------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-------------------------------------------------------- |
| imageUrls              | array                                                                                  | yes      | Image Source                                                                                                                                                                                                                         |                                                           |
| enableImageZoom        | boolean                                                                                | no       | Enable image zoom                                                                                                                                                                                                                    | `true`                                                    |
| onShowModal            | function<br><br>`(content?: JSX.Element) => void`                                      | no       | The callback for show modal                                                                                                                                                                                                          | `() => {}`                                                |
| onCancel               | function<br><br>`() => void`                                                           | no       | The callback for cancel modal                                                                                                                                                                                                        | `() => {}`                                                |
| flipThreshold          | number                                                                                 | no       | Swipe threshold of the next page                                                                                                                                                                                                     | `80`                                                      |
| maxOverflow            | number                                                                                 | no       | The X position maximum, that current page can slide to the next page                                                                                                                                                                 | `300`                                                     |
| index                  | number                                                                                 | no       | Init index of images                                                                                                                                                                                                                 | `0`                                                       |
| failImageSource        | string, object<br><br>`{url: string}`                                                  | no       | placeholder for fail                                                                                                                                                                                                                 | `''`                                                      |
| loadingRender          | function<br><br>`() => React.ReactElement<any>`                                        | no       | placeholder for loading                                                                                                                                                                                                              | `() => null`                                              |
| onSaveToCamera         | function<br><br>`(index?: number => void`                                              | no       | The callback for save to camera                                                                                                                                                                                                      | `() => {}`                                                |
| onChange               | function<br><br>`(index?: number => void`                                              | no       | When the image changed                                                                                                                                                                                                               | `() => {}`                                                |
| onMove                 | ( position: [IOnMove](https://github.com/ascoders/react-native-image-zoom/blob/master/src/image-zoom/image-zoom.type.ts) )=>void | reports movement position data (helpful to build overlays)                                                                                                            | ()=> {}      |
| saveToLocalByLongPress | boolean                                                                                | no       | Enable save to camera when long press                                                                                                                                                                                                | `true`                                                    |
| onClick                | function<br><br>`(onCancel?: function) => void`                                        | no       | Onclick                                                                                                                                                                                                                              | `(onCancel) => {onCancel()}`                              |
| onDoubleClick          | function<br><br>`(onCancel?: function) => void`                                        | no       | OnDoubleClick                                                                                                                                                                                                                        | `(onCancel) => {onCancel()}`                              |
| onSave                 | function<br><br>`(url: string) => void`                                                | no       | The picture is saved to the local method, if you write this method will not call the system default method for Android does not support saveToCameraRoll remote picture, you can call this callback in Android call native interface |                                                           |
| renderHeader           | function<br><br>`(currentIndex?: number) => React.ReactElement<any>`                   | no       | Custom header                                                                                                                                                                                                                        | `() => null`                                              |
| renderFooter           | function<br><br>`(currentIndex?: number) => React.ReactElement<any>`                   | no       | Custom footer                                                                                                                                                                                                                        | `() => null`                                              |
| renderIndicator        | function<br><br>`(currentIndex?: number, allSize?) => React.ReactElement<any>`: number | no       | Custom indicator                                                                                                                                                                                                                     | `(currentIndex, allSize) => currentIndex + "/" + allSize` |
| renderImage            | function<br><br>`(props: any) => React.ReactElement<any>`                              | no       | Custom image component                                                                                                                                                                                                               | `(props) => <Image {...props} />`                         |
| renderArrowLeft        | function<br><br>`() => React.ReactElement<any>`                                        | no       | Custom left arrow                                                                                                                                                                                                                    | `() => null`                                              |
| renderArrowRight       | function<br><br>`() => React.ReactElement<any>`                                        | no       | Custom right arrow                                                                                                                                                                                                                   | `() => null`                                              |
| onSwipeDown            | function<br><br>`() => void`                                                           | no       | Callback for swipe down                                                                                                                                                                                                              | `() => null`                                              |
| footerContainerStyle   | object<br><br>`{someStyle: someValue}`                                                 | no       | custom style props for container that will be holding your footer that you pass                                                                                                                                                      | `bottom: 0, position: "absolute", zIndex: 9999`           |
| backgroundColor        | string<br><br>`white`                                                                  | no       | Component background color                                                                                                                                                                                                           | `black`                                                   |
| enableSwipeDown        | boolean                                                                                | no       | Enable swipe down to close image viewer. When swipe down, will trigger onCancel.                                                                                                                                                     | `false`                                                   |
| swipeDownThreshold     | number                                                                                 | no       | Threshold for firing swipe down function                                                                                                                                                                                             |                                                           |
| doubleClickInterval    | number                                                                                 | no       | Double click interval.                                                                                                                                                                                                               |                                                           |
| pageAnimateTime        | number                                                                                 | no       | Set the animation time for page flipping.                                                                                                                                                                                            | `100`                                                     |
| enablePreload          | boolean                                                                                | no       | Preload the next image                                                                                                                                                                                                               | `false`                                                   |
| useNativeDriver        | boolean                                                                                | no       | Whether to animate using [`useNativeDriver`](https://reactnative.dev/docs/animations#using-the-native-driver)                                                                                                                        | `false`                                                   |
| menus                  | function<br><br>`({cancel,saveToLocal}) => React.ReactElement<any>`                    | no       | Custom menus, with 2 methods:`cancel` to hide menus and `saveToLocal` to save image to camera
| menuContext            | object<br><br>`{someKey: someValue}`                                                   | no       | Custom menu context.                              | `{ saveToLocal: 'save to the album', cancel: 'cancel' }`
## Development pattern

### Step 1, run TS listener

After clone this repo, then:

```bash
npm install
npm start
```

### Step 2, run demo

```bash
cd demo
npm install
npm start
```

Then, scan the QR, use your [expo app](https://expo.io./).

### Dependence

Depend on `react-native-image-pan-zoom`: https://github.com/ascoders/react-native-image-zoom


History.md

1.1.0 / 2023-08-16
==================

  * feat: 支持修改 imageUrls
  * fixed: 解决 imageUrls 变化后会导致崩溃的bug
  * 3.0.1
  * Update react-native-image-pan-zoom to 2.1.12
  * Merge pull request #394 from pxpeterxu/master
  * Update documentation
  * Allow specifying whether to use useNativeDriver
  * release 3.0.0
  * Merge pull request #389 from radeno/patch-1
  * Do not break compatibility with existed implementation
  * Add compatibility with RN 0.62
  * deploy 2.2.27
  * Merge pull request #338 from alexiri/scale
  * Merge pull request #358 from ArtemKolichenkov/update-deprecated-lifecycle-methods
  * Merge pull request #356 from ArtemKolichenkov/upgrade-expo-sdk-for-demo
  * Upgrade Expo SDK to 35.0.0 for demo app
  * componentWillReceiveProps fix
  * Update componentWillMount to componentDidMount
  * Expose minScale and maxScale
  * Merge pull request #324 from DISKONEKTeD/master
  * out of bounds index fix: renderFooter currentIndex
  * Merge pull request #316 from fgreinus/patch-1
  * Merge pull request #315 from shawerestart/master
  * Fix local image loading example in Readme
  * [update] update doc
  * fix typo error
  * fix type error
  * deploy 2.2.26
  * Merge pull request #307 from React-Sextant/master
  * Add custom menus component
  * Prohibit any panResponder when longPress to show menu
  * Merge pull request #305 from Ovi/patch-2
  * renderFooter must return index
  * Merge pull request #303 from Ovi/patch-1
  * Update ReadMe.md
  * Merge pull request #302 from pt7892/master
  * Added onMove prop to ImageZoom
  * Merge pull request #298 from julianapinheiro/fix-fail-status
  * fix fail status never saved
  * Update README.md
  * deploy 2.2.25
  * Merge pull request #279 from Jimbomaniak/fix/centered_loader
  * fix show loader in the center of screen
  * Merge pull request #270 from corygibbons/readme
  * Minor README.md formatting
  * deploy 2.2.24
  * Merge pull request #264 from wack17s/swipeDownThreshold
  * fix readme
  * add swipeDownThreshold prop
  * deploy 2.2.23
  * Merge pull request #252 from LDty/master
  * fix z-index error in renderIndicator
  * deploy 2.2.22
  * Merge pull request #250 from nbolender/fix-header-touches
  * Add negative zIndex to indicator component so that headerComponent can receive touches
  * deploy 2.2.21
  * Merge pull request #245 from sanqi/master
  * fix enablePreload
  * add pageAnimateTime
  * Merge branch 'master' of https://github.com/ascoders/react-native-image-viewer
  * deploy 2.2.20
  * Merge pull request #244 from alvinmatias69/master
  * Update expo and RN dependencies on demo
  * Merge pull request #242 from sanqi/master
  * Update image-viewer.component.tsx
  * Update image-viewer.type.ts
  * fix
  * add Preload
  * Support to set the source image width and height
  * Merge pull request #241 from tomhicks/patch-2
  * Add declaration config to correct file
  * Remove wrong configuration
  * deploy 2.2.19
  * Merge pull request #239 from mei-mu/master
  * Merge pull request #240 from tomhicks/patch-1
  * Reference external declaration file
  * Generate type declaration file
  * add reset image
  * add reset image
  * support doubleClickInterval
  * deploay 2.2.17
  * support enableImageZoom
  * upgrade deps
  * release 2.2.14
  * Merge pull request #216 from noambonnie/bugfix-rtl-index
  * Merge pull request #217 from noambonnie/remove-log-message
  * Remove log message
  * RTL Bug fix for initial index
  * Merge pull request #209 from huangming1994/master
  * update README
  * upgrade deps
  * update
  * fix goNext
  * release 2.2.10
  * support enableSwipeDown
  * release 2.2.8
  * release 2.2.7
  * release 2.2.7
  * Merge pull request #202 from mickaelw/fix/missing-min-and-max-scale-props
  * Fix: Mission required props from RN image pan zoom package
  * fix #182
  * Merge branch 'master' of https://github.com/ascoders/react-native-image-viewer
  * fit ts2.9
  * Merge pull request #190 from nicomontanari98/patch-1
  * Removed unused param
  * fix first image not show
  * fix: TypeError for Object.assigng
  * deploy 2.2.3
  * Merge branch 'master' of https://github.com/ascoders/react-native-image-viewer
  * deploy 2.2.2; merged #178, #180
  * Merge pull request #178 from bradsolves/master
  * Merge pull request #180 from bolandrm/master
  * allow user to specify custom image renderer
  * Fixed typo
  * Added Rotation
  * support RTL
  * optimizes the performance of show pictures
  * Merge branch 'master' of https://github.com/ascoders/react-native-image-viewer
  * support override url option
  * pass merged props to image
  * Merge pull request #166 from aterribili/patch-1
  * remove invalid changes
  * Add backgroundColor to documentation
  * Release 2.1.2, fix #161
  * update readme
  * Release 2.1.0, support swipe down
  * rollback swiperDown
  * Merge pull request #154 from ShayanJavadi/master
  * fixed renderfooter not rendering properly by adding a footer container. updated docs to reflect changes
  * fixed bug where no imageurl is passed
  * Merge pull request #153 from ShayanJavadi/master
  * updated docs
  * made animation become smoohter
  * added swipe down feature
  * Merge pull request #150 from beilunyang/master
  * chore:remove unused variables
  * fix #142
  * update readme
  * fix: performance problems
  * Merge pull request #131 from RobertSheaO/master
  * Add examples to prop table
  * Transcribe document image to markdown table
  * Update readme.md
  * Merge pull request #125 from jakewtaylor/master
  * Remove unused dependency
  * Switch `transmit-transparently` dependency
  * fix: double click bug
  * format code
  * restructure
  * Merge pull request #120 from andrew/patch-1
  * Update license field in package.json to match LICENSE file
  * deploy 2.0.16
  * Merge pull request #113 from timothystewart6/patch-1
  * fix(package): 2.0.14 does not exist
  * upgrade: deps
  * fix #101
  * deploy 2.0.13
  * Merge pull request #106 from yunkaiwang/master
  * add image as a parameter to the onLongPress function

src/image-viewer.type.ts
import * as React from 'react';
import { Image, ImageURISource, Text, View, ViewStyle } from 'react-native';
import { simpleStyle } from './image-viewer.style';

interface IOnMove {
  type: string;
  positionX: number;
  positionY: number;
  scale: number;
  zoomCurrentDistance: number;
}

export class Props {
  /**
   * 是否显示
   */
  public show?: boolean = false;

  /**
   * 图片数组
   */
  public imageUrls: IImageInfo[] = [];

  /**
   * 滑动到下一页的X阈值
   */
  public flipThreshold?: number = 80;

  /**
   * 当前页能滑到下一页X位置最大值
   */
  public maxOverflow?: number = 300;

  /**
   * 初始显示第几张图
   */
  public index?: number = 0;

  /**
   * 加载失败的图
   */
  public failImageSource?: IImageInfo = undefined;

  /**
   * 背景颜色
   */
  public backgroundColor?: string = 'black';

  /**
   * style props for the footer container
   */
  public footerContainerStyle?: object = {};

  /**
   * Menu Context Values
   */
  public menuContext?: any = { saveToLocal: 'save to the album', cancel: 'cancel' };

  /**
   * 是否开启长按保存到本地的功能
   */
  public saveToLocalByLongPress?: boolean = true;

  /**
   * 是否允许缩放图片
   */
  public enableImageZoom?: boolean = true;

  public style?: ViewStyle = {};

  /**
   * Enable swipe down to close image viewer.
   * When swipe down, will trigger onCancel.
   */
  public enableSwipeDown?: boolean = false;

  /**
   * threshold for firing swipe down function
   */
  public swipeDownThreshold?: number;

  public doubleClickInterval?: number;

  /**
   * Min and Max scale for zooming
   */
  public minScale?: number;

  public maxScale?: number;

  /**
   * 是否预加载图片
   */
  public enablePreload?: boolean = false;

  /**
   * 翻页时的动画时间
   */
  public pageAnimateTime?: number = 100;

  /** 
   * 是否启用原生动画驱动
   * Whether to use the native code to perform animations.
   */
  public useNativeDriver?: boolean = false;

  /**
   * 长按图片的回调
   */
  public onLongPress?: (image?: IImageInfo) => void = () => {
    //
  };

  /**
   * 单击回调
   */
  public onClick?: (close?: () => any, currentShowIndex?: number) => void = () => {
    //
  };

  /**
   * 双击回调
   */
  public onDoubleClick?: (close?: () => any) => void = () => {
    //
  };

  /**
   * 图片保存到本地方法，如果写了这个方法，就不会调取系统默认方法
   * 针对安卓不支持 saveToCameraRoll 远程图片，可以在安卓调用此回调，调用安卓原生接口
   */
  public onSave?: (url: string) => void = () => {
    //
  };

  public onMove?: (position?: IOnMove) => void = () => {
    //
  };

  /**
   * 自定义头部
   */
  public renderHeader?: (currentIndex?: number) => React.ReactElement<any> = () => {
    return null as any;
  };

  /**
   * 自定义尾部
   */
  public renderFooter?: (currentIndex: number) => React.ReactElement<any> = () => {
    return null as any;
  };

  /**
   * 自定义计时器
   */
  public renderIndicator?: (currentIndex?: number, allSize?: number) => React.ReactElement<any> = (
    currentIndex?: number,
    allSize?: number
  ) => {
    return React.createElement(
      View,
      { style: simpleStyle.count },
      React.createElement(Text, { style: simpleStyle.countText }, currentIndex + '/' + allSize)
    );
  };

  /**
   * Render image component
   */
  public renderImage?: (props: any) => React.ReactElement<any> = (props: any) => {
    return React.createElement(Image, props);
  };

  /**
   * 自定义左翻页按钮
   */
  public renderArrowLeft?: () => React.ReactElement<any> = () => {
    return null as any;
  };

  /**
   * 自定义右翻页按钮
   */
  public renderArrowRight?: () => React.ReactElement<any> = () => {
    return null as any;
  };

  /**
   * 弹出大图的回调
   */
  public onShowModal?: (content?: any) => void = () => {
    //
  };

  /**
   * 取消看图的回调
   */
  public onCancel?: () => void = () => {
    //
  };

  /**
   * function that fires when user swipes down
   */
  public onSwipeDown?: () => void = () => {
    //
  };

  /**
   * 渲染loading元素
   */
  public loadingRender?: () => React.ReactElement<any> = () => {
    return null as any;
  };

  /**
   * 保存到相册的回调
   */
  public onSaveToCamera?: (index?: number) => void = () => {
    //
  };

  /**
   * 当图片切换时触发
   */
  public onChange?: (index?: number) => void = () => {
    //
  };

  public menus?: ({ cancel, saveToLocal }: any) => React.ReactElement<any>;

  public handleHorizontalOuterRangeOffset?: (offsetX: number) => {
    // 
  };
}

export class State {
  /**
   * 是否显示
   */
  public show?: boolean = false;

  /**
   * 当前显示第几个
   */
  public currentShowIndex?: number = 0;

  /**
   * Used to detect if parent component applied new index prop
   */
  public prevIndexProp?: number = 0;

  /**
   * 图片拉取是否完毕了
   */
  public imageLoaded?: boolean = false;

  /**
   * 图片长宽列表
   */
  public imageSizes?: IImageSize[] = [];

  /**
   * 是否出现功能菜单
   */
  public isShowMenu?: boolean = false;
}

export interface IImageInfo {
  url: string;
  /**
   * 没有的话会自动拉取
   */
  width?: number;
  /**
   * 没有的话会自动拉取
   */
  height?: number;
  /**
   * 图片字节大小(kb为单位)
   */
  sizeKb?: number;
  /**
   * 原图字节大小(kb为单位)
   * 如果设置了这个字段,并且有原图url,则显示查看原图按钮
   */
  originSizeKb?: number;
  /**
   * 原图url地址
   */
  originUrl?: string;
  /**
   * Pass to image props
   */
  props?: any;
  /**
   * 初始是否不超高 TODO:
   */
  freeHeight?: boolean;
  /**
   * 初始是否不超高 TODO:
   */
  freeWidth?: boolean;
}

export interface IImageSize {
  width: number;
  height: number;
  // 图片加载状态
  status: 'loading' | 'success' | 'fail';
}
